xclim/core/calendar.py

```diff
@@ -453,10 +459,40 @@ def percentile_doy(
     #  This is supported by DataArray.quantile, but not by groupby.reduce.
     rr = arr.rolling(min_periods=1, center=True, time=window).construct("window")
 
-    # Create empty percentile array
-    g = rr.groupby("time.dayofyear")
+    # rolling automatically adds some nans to the windows on the time-series edges
+    # Optionally fill nans in first and last n=floor(window/2) time steps with average window values for these
+    if fill_ends:
+        nstep = floor(window / 2)
```

Comment:
This is not my proudest piece of work... is there a more elegant way to pad the data with the mean

Reply:
This is assuming we _always_ want to skip NaNs.

Reply:
I was trying to avoid the case of filling all nans... These ones are a simple artifact resulting from the n=5 window construction.  Not sure if is reflected in the tests for `calendar` but in an application with 30 years of data :  dayofyear[0] there will have two nans tacked on to the beginning of window dimension but only for the first year of the `rr` variable.  My idea was to replace nan with only the mean of the 3 valid days of single year/dayofyear window and not the mean of all years.   Originally thought of just a bfill() for the begginning of the timeseries and ffill() for the end but the mean seemed like a better choice.

Reply:
In a case where there are other nans present in the data ... i.e. actual missing day in the original timeseries the masking will catch it and use xr.quantile(skipna=True) but again only where it really needs it.  Overall making things much faster as long as the data is relatively complete

Reply:
In a 3d case (time, lat, lon) I think this means that even with "perfect data" (i.e. no nans) there is minimally `n=(floor(window/2) * 2) * len(ds.lon) * len(ds.lat)` cases where the slower `xr.quantile(skipna=True)` call is used.  As spatial domain gets bigger the number of these cases grows quickly and there will likely be increased difference between `fill_ends=True`  and `fill_ends=False`